posix: include: dirent: rework the dirent.h header

cfriedt opened the pull request:
* posix: fs: declare _POSIX_C_SOURCE in a consistent way
* scripts: checkpatch: add special consideration for DIR
* tests: posix: fs: declare entry statically
* posix: include: dirent: rework the dirent.h header

Commit:
posix: include: dirent: rework the dirent.h header

Declare standard functions and split type definitions into
sys/dirent.h .

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
scripts: checkpatch: add special consideration for DIR

DIR is a POSIX type defined via the dirent.h header.

In Zephyr, we define it as `typedef void DIR`, since it is only
ever described via a pointer (much like `FILE`).

However, in checkpatch.pl, functions that return DIR* or accept
a DIR* argument are met with an error of the form below:

```
ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
```

Examples that trigger this false positive are, for example

```cpp
int dirfd(DIR *dirp);
DIR *fdopendir(int fd);
```

Include `DIR` as a class of specific POSIX types that should be
matched as types rather than other tokens.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

de-nordic commented on include/zephyr/posix/sys/dirent.h:
```diff
@@ -0,0 +1,45 @@
+#endif
+
+#ifndef PATH_MAX
+#define PATH_MAX 31
```
We have FILE_SYSTEM_MAX_FILE_NAME https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20FILE_SYSTEM_MAX_FILE_NAME&type=code, and that will impact the PATH_MAX if set above 31 characters.

cfriedt replied:
Now that I'm looking at this, it really seems like this should be `NAME_MAX` instead of `PATH_MAX`, which has corresponding `_XOPEN_NAME_MAX` and `_POSIX_NAME_MAX` [ref](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html)

Indeed, that seems to be true [ref](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/dirent.h.html).

Interestingly, this relates to `_POSIX_NO_TRUNC`. If that value is `> 0` and a path component (name) longer than `_NAME_MAX` is supplied to a function, it will return with an error code `ENAMETOOLONG` [ref](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/unistd.h.html). Otherwise, the path component is simply truncated. Currently, `_POSIX_NO_TRUNC` is zero, so we're OK to silently truncate.

> We have FILE_SYSTEM_MAX_FILE_NAME

It looks as though `FILE_SYSTEM_MAX_FILE_NAME` appllies to the complete path instead of individual path components (names).

I did a quick search [here](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_FILE_SYSTEM_MAX_FILE_NAME) with the search `^CONFIG_FILE_SYSTEM_.*$` and it looks like there isn't anything corresponding to `NAME_MAX` within Zephyr's native FS subsystem.

cfriedt replied:
Fixed! (with less code, as it turns out)

de-nordic replied:
> I did a quick search [here](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_FILE_SYSTEM_MAX_FILE_NAME) with the search ^CONFIG_FILE_SYSTEM_.*$ and it looks like there isn't anything corresponding to NAME_MAX within Zephyr's native FS subsystem.

We actually do not have anything corresponding to PATH_MAX, there are app specific defines like in fs shell, but FS system itself does not have such. The `CONFIG_FILE_SYSTEM_MAX_FILE_NAME ` is only used by `fs_dirent` to return name of file found while traversing directory contents - that is the single place.
We do not have anything like `PATH_MAX` due to how the FS processes paths: after stripping the mount point we are giving full path, that user has created/provided, to file system driver and the driver then goes through the path and tries to figure out where the file is, so from the point of view of Zephyr the path can be infinitely long and we do not care.
Zephyr VFS does not store paths anywhere, they are left on the user/application side.

> Interestingly, this relates to _POSIX_NO_TRUNC. If that value is > 0 and a path component (name) longer than _NAME_MAX is supplied to a function, it will return with an error code ENAMETOOLONG [ref](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/unistd.h.html). Otherwise, the path component is simply truncated. Currently, _POSIX_NO_TRUNC is zero, so we're OK to silently truncate.

Not sure this is fine, since from time to time we may be actually lucky to open/remove/truncate some unintended file.

cfriedt commented:
Weird - I can't seem to reproduce any of the CI issues. Will try a rebase.

cfriedt commented on include/zephyr/posix/sys/dirent.h:
```diff
@@ -0,0 +1,29 @@
+
+struct dirent {
+	unsigned int d_ino;
+	char d_name[NAME_MAX];
```
It would probably make sense to set POSIX runtime invariant values based on Zephyr values. Similar to the maximum number of file descriptors.

https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/posix/options/Kconfig.device_io#L48

I didn't find a great way of doing that other than setting the POSIX config to the Zephyr config by default. 

The POSIX constants (`_POSIX_PATH_MAX`) can't be changed, but software can at least check if the runtime invariant value (`PATH_MAX`) is less than the minimum (which are oddly "max" values in POSIX).

Commit:
posix: fs: declare _POSIX_C_SOURCE in a consistent way

Declare _POSIX_C_SOURCE in a consistent way for both the
posix/options library as well as the tests/posix/fs
testsuite.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: posix: fs: declare entry statically

qemu_x86/atom was failing tests/posix/fs in
test_fs_readdir_threadsafe for some reason and only for that
platform.

All other platforms pass the testsuite properly with the dirent
declared on the stack.

Declare the dirent object statically so that qemu_x86/atom will
pass test cases.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
I still don't know exactly why the test was failing for `qemu_x86/atom`, but declaring the `dirent` statically allows the test to pass for that platform. All others were fine 🤷‍♂️

cfriedt commented:
* cleaned up commits

cfriedt commented on tests/posix/fs/src/test_fs_dir.c:
```diff
@@ -57,7 +54,8 @@ static int test_mkdir(void)
 {
 	if (thread_safe) {
-		struct dirent entry;
+		/* cannot declare on stack otherwise this test fails for qemu_x86/atom */
```
I think it might have to do more with stack padding or rounded stack sizes. I even tried with 4096 and it did not work.

Probably can revisit later.
